### ISSUE TITLE
Creating subquery while get count from relation with orderBy

### ISSUE THREAD

**Issue by lebedevsky** [yiisoft/yii2#14132]
When in relation use orderBy and next get count from \yii\db\ActiveQuery, it will be use subquery to select count, as fix may be rewrite orderBy by set empty array.

### What steps will reproduce the problem?

```
class Category {
...
    public function getItems()
    {
        return $this->hasMany(Item::className(), ['category_id' => 'id'])
                ->orderBy(['date' => SORT_DESC]);
    }
...
}
```

`$category->getItems()->count()`

### What is the expected result?

`SELECT COUNT(*) FROM `items` WHERE `category_id` = ? ORDER BY `date` DESC`

### What do you get instead?

`SELECT COUNT(*) FROM (SELECT * FROM `items` WHERE `category_id`= ? ORDER BY `date` DESC) `c``

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.11.2
| PHP version      | 7.1.4
| Operating system | centos 7

**Comment by cebe** [comment]
what is the problem? There should be no difference in the result as far as I see.

**Comment by cebe** [comment]
related to #13846

**Comment by kakalos12** [comment]
Hi there, I'm having a problem with the count and subquery too. 
In my database, I have 2 lists, a subscriber can subscribe to 2 lists. so there is a junction table. And in the query I use the code like this. 
`$subquery = (new Query())->select([
            new \yii\db\Expression('1')
        ])
            ->from("subscriber_list")
            ->where([
            'list_id' => $id
        ])
            ->andWhere("subscribers.id = subscriber_list.subscriber_id");
        $subscribers = (new Query())->select("subscribers.*, subscriber_list.subscribed_at, group_concat(subscriber_list.list_id) as lists")
            ->from("subscribers")
            ->innerJoin("subscriber_list", "subscriber_list.subscriber_id = subscribers.id")
            ->where([
            'exists',
            $subquery
        ])
            ->andFilterWhere([
            'or',
            [
                'like',
                'name',
                $query
            ],
            [
                'like',
                'email',
                $query
            ]
        ]);`

Expected Pagination-total count : 1 
Result : 2.

**Comment by cebe** [comment]
@lebedevsky this should have been fixed in #13846

**Comment by combinator** [comment]
I think the fix may be problematic. I'll illustrate my point with the following scenario : 
1. I have a table with more than 3 million rows. 
2. Among these rows , some users may have 10 records, others , say 700,000 records. 
3. the goal is to display the most recent records for a particular user . Since some users may have very HIGH number of records, i've decided to max number of records to be displayed to 1000 (using LIMIT)  (and the result set will be paginated) . 
Here comes the problem . Suppose that userA has 700+ k records. 
If I run this query  (example) :
`SELECT COUNT(*) FROM items WHERE user=userA_Id   ORDER BY date DESC LIMIT 1000`
Then , the COUNT will have to count up to 700+ k records for userA which , in my cases , takes **more than 2000ms**.
Note here that the LIMIT 1000 has not impact to the COUNT since COUNT will still perform a full scan for UserA_ID

However, this command : 
`SELECT COUNT(*) FROM (SELECT * FROM items WHERE userId= userA_Id  ORDER BY date DESC LIMIT 1000 ) as T `
takes about ~**2ms**.

As you can see, there's are differences in SQL syntax and potential performance consequences between the aforementioned commands. 
Therefore I think that the previous change should be reverted or at least that there should be a function for COUNT to count the result of a subquery (which may be small subset as dictated by the LIMIT option 
 instead of making COUNT count ALL records . 
I've been struggling to find a *Clean* way of using subqueries (as per example #2) to no avail.
How can I achieve the second scenario using Yii2 ? 
Would appreciate your feedback on this matter.

> Yii version : 2.0.14.1 

Thanks

**Comment by SilverFire** [comment]
TLDR: There is not silver bullet. The change you suggest will improve your query performance, but will slow down other queries. You should fix it on your side.

____

Sometimes a query performance can be improved (or worsen) by changing something really not obvious in SQL. It happens because the SQL is only an abstract language. SQL optimizer of your DBMS builds an execution plan and the magic that happens there sometimes is not predictable.

I know a good example from my daily life with PgSQL. 

```sql
-- Find requests for clients of seller
SELECT * FROM requests WHERE client_id IN (SELECT id FROM client WHERE seller_id = 8418243);
-- Find clients that have requests
SELECT * FROM client WHERE id IN (SELECT DISTINCT client_id FROM requests);
```

vs

```sql
-- Find requests for clients of seller
SELECT * FROM requests WHERE client_id = ANY(array(SELECT id FROM client WHERE seller_id = 8418243));
-- Find clients that have requests
SELECT * FROM client WHERE id = ANY(array(SELECT DISTINCT client_id FROM requests));
```

Which one should be faster, considering every column included in this query is indexed? :) Think before reading the answer.

<details><summary> :boom: Answer  </summary><p>
The correct answer is "<i>it depends</i>". It depends on how many records will be returned in subquery (how many clients of the seller exist), on how many records are now in the `requests` table, how spread are they between clients. And with every single set of "<i>it depends</i>" the execution plan got changed and performance as well.
</p></details>
